File: src/stocks/src/handlers/topic.cs
using Confluent.Kafka;
using System.Net;
using System.Text.Json;
using System.Collections.Generic;

namespace Stocks
{
    // TODO: use dotenv
    const KAFKA_BOOTSTRAP_SERVER = "";

    class TopicHandler
    {
        private Producer producer;
        private Consumer consumer;

        public TopicHandler()
        {
            this.producer = new ProducerBuilder<Null, string>(new ProducerConfig
            {
                BootstrapServers = KAFKA_BOOTSTRAP_SERVER,
            }).Build();

            this.consumer = new ConsumerBuilder<Ignore, string>(new ConsumerConfig
            {
                BootstrapServers = KAFKA_BOOTSTRAP_SERVER,
            }).Build();
        }

        private void deserialize(Dictionary<string, string> raw)
        {
            // TODO: refactor parse into deserialize
            return ShopOrderRequest.parse(raw);
        }

        private void decode()
        {
            return deserialize(JsonSerializer.Deserialize<Dictionary<string, string>>(raw))
        }

        public void read(string topic, Func<ShopOrderRequest> onMessage)
        {
            consumer.Subscribe(topic);

            while (!cancelled)
            {
                var consumeResult = consumer.Consume(cancellationToken);
                onMessage(decode((string)consumeResult.value));
            }

            consumer.Close();
        }

        private void serialize(ShopOrderRequest shopOrderRequest)
        {
            return shopOrderRequest.serialize();
        }

        private void encode(ShopOrderRequest shopOrderRequest)
        {
            return (string)serialize(shopOrderRequest);
        }

        public void send(string topic, ShopOrderRequest shopOrderRequest)
        {
            var result = await producer.ProduceAsync(topic, new Message<Null, string> { Value = encode(shopOrderRequest) });
        }
    }
}

File: src/invoices/src/handlers/topic.py
import json
from typing import Dict

from kafka import KafkaConsumer, KafkaProducer
from src.models.shopOrderRequest import ShopOrderRequest

# TODO: use dotenv configuration
KAFKA_BROKER = ""


class TopicHandler:
    consumers: Dict[str, KafkaConsumer]
    producer: KafkaProducer

    def deserialize(self, payload):
        return ShopOrderRequest.parse(json.loads(payload))

    def decode(self, message):
        return self.deserialize(str(message).decode(encoding="UTF-8"))

    def read(self, topic: str):
        if topic not in self.consumers:
            self.consumers[topic] = KafkaConsumer(
                topic,
                bootstrap_servers=KAFKA_BROKER
            )

        for message in self.consumers[topic]:
            yield self.decode(message)

    def serialize(self, payload: ShopOrderRequest):
        return str(payload)

    def encode(self, message):
        return self.serialize(message).encode(encoding="UTF-8")

    def send(self, topic, message):
        if not self.producer:
            self.producer = KafkaProducer(bootstrap_servers=KAFKA_BROKER)

        return self.producer.send(topic, value=self.encode(message)).get(10)
